telemetry-proxy/telemetry_proxy.go

```diff
@@ -20,9 +20,100 @@ import (
 	"google.golang.org/protobuf/encoding/protojson"
 )
 
-// connections stores active WebSocket connections.
-var connectionsMutex sync.RWMutex
-var connections = make(map[*websocket.Conn]bool)
+type WebClients struct {
+	ctx      context.Context
+	upgrader *websocket.Upgrader
+	mu       sync.Mutex
+	clients  map[*websocket.Conn]bool
+}
+
+var webClients WebClients
+
+// Start will initialize the struct and listen for connections.
+func (w *WebClients) Start(ctx context.Context) error {
+	w.ctx = ctx
+	w.upgrader = &websocket.Upgrader{
+		ReadBufferSize:  1024,
+		WriteBufferSize: 1024,
+		CheckOrigin: func(r *http.Request) bool {
+			return true // Allow all origins for now (we might want to restrict this in production).
+		},
+	}
+	w.clients = map[*websocket.Conn]bool{}
+
+	// Start the WebSocket Server on port 8080.
+	http.HandleFunc("/ws", func(rw http.ResponseWriter, r *http.Request) {
+		w.Handle(rw, r)
+	})
+	err := http.ListenAndServe(":8080", nil)
+	return err
+}
+
+// Send writes data over all WebSocket connections in w.clients.
+func (w *WebClients) Send(data []byte) {
+	// Iterate through all active connections and send the data
+	w.mu.Lock()
+	defer w.mu.Unlock()
+	for c := range w.clients {
+		if err := c.WriteMessage(websocket.TextMessage, data); err != nil {
+			log.Println("Error writing message to WebSocket client:", err)
+			// Remove the connection from the map if there's an error
+			delete(w.clients, c)
+			// Close the connection
+			c.Close()
+		}
+	}
+}
+
+// Handle upgrades the connection, adds a client to the list, and starts serving it.
+func (w *WebClients) Handle(rw http.ResponseWriter, r *http.Request) {
+	// Upgrade the HTTP connection to a WebSocket connection.
+	conn, err := w.upgrader.Upgrade(rw, r, nil)
+	if err != nil {
+		log.Println("Error upgrading connection:", err)
+		return
+	}
+
+	// Add the new connection to the map
+	w.mu.Lock()
+	w.clients[conn] = true
+	w.mu.Unlock()
+
+	conn.SetCloseHandler(func(code int, text string) error {
+		w.mu.Lock()
+		defer w.mu.Unlock()
+		delete(w.clients, conn)
+		return nil
+	})
+
+	go w.Serve(conn)
+}
+
+// Serve gets called within a goroutine from Handle.
+func (w *WebClients) Serve(conn *websocket.Conn) {
+	for {
+		select {
+		case <-w.ctx.Done(): // Handle context cancellation (server shutdown)
+			conn.WriteControl(websocket.CloseMessage, websocket.FormatCloseMessage(websocket.CloseNormalClosure, "Server shutting down"), time.Now().Add(5*time.Second))
+			return // Exit the goroutine
+
+		default:
+			messageType, p, err := conn.ReadMessage()
+			if err != nil {
+				log.Println("Error reading message:", err)
+				return
+			}
+			log.Printf("Received message: %s\n", p)
+
+			// Process Message
+			w.mu.Lock()
```

Comment:
We write messages in two places (broadcasting telemetry and processing requests), and only one writer can be alive at a time. Without a mutex lock, I was running into race conditions. A future optimization could be to make per-connection mutexes and use those, but for now (and likely forever), this will suffice and is simpler